src/hmac.rs

```diff
@@ -0,0 +1,326 @@
+// Bitcoin Hashes Library
+// Written in 2018 by
+//   Andrew Poelstra <apoelstra@wpsoftware.net>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+// This module is largely copied from the rust-crypto ripemd.rs file;
+// while rust-crypto is licensed under Apache, that file specifically
+// was written entirely by Andrew Poelstra, who is re-licensing its
+// contents here as CC0.
+
+//! # HMAC support
+
+use std::{fmt, io, ops};
+
+use {Error, Hash};
+
+/// A hash computed from a RFC 2104 HMAC. Parameterized by the underlying hash function.
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub struct Hmac<T: Hash>(T);
+
+/// Pair of underyling hash engines, used for the inner and outer hash of HMAC
+#[derive(Clone)]
+pub struct HmacEngine<T: Hash> {
+    iengine: T::Engine,
+    oengine: T::Engine,
+}
+
+impl<T: Hash> HmacEngine<T> {
+    /// Construct a new keyed HMAC with the given key
+    pub fn new(key: &[u8]) -> HmacEngine<T> {
+        debug_assert!(T::block_size() <= 128);
+
+        let mut ipad = [0x36u8; 128];
+        let mut opad = [0x5cu8; 128];
+        let mut ret = HmacEngine {
+            iengine: T::engine(),
+            oengine: T::engine(),
+        };
+
+        if key.len() > T::block_size() {
+            let hash = T::hash(key);
+            for (i, byte) in hash[..].iter().enumerate() {
+                ipad[i] ^= *byte;
+                opad[i] ^= *byte;
+            }
+        } else {
+            for (i, byte) in key.iter().enumerate() {
```

Comment:
Can you propose different code? I think it'd be less readable and less efficient to be switching inside a loop, and we'd save only one or two lines of code.

Reply:
Ah, you mean to just assign either `key` or `T::hash(key)` to the `hash` and then do a single loop on the result. I can't see how to do this because `key` and `T::hash(key)` have different types, and converting them to slices too early results in borrowck errors.

Reply:
Right, I didn't think of that at first. But I found a solution using `Option`:

```rust
let maybe_key_hash = if key.len() > T::block_size() {
    Some(T::hash(key))
} else {
    None
};
let derived_key = maybe_key_hash.as_ref().map(|kh| &kh[..]).unwrap_or(key);

for (b_i, b_h) in ipad.iter_mut().zip(derived_key) {
    *b_i ^= *b_h;
}
for (b_o, b_h) in opad.iter_mut().zip(derived_key) {
    *b_o ^= *b_h;
}
```

Reply:
I know that it's not beautiful but at least it avoids redundancy and isn't longer.

Reply:
I think the redundant code is clearer, even though it's redundant, because there are only 2 blocks and their similarity is visually immediate.

Reply:
Agree